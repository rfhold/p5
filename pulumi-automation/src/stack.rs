#![allow(dead_code)]

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use tokio::sync::mpsc::Sender;

use crate::event::EngineEvent;

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceState {
    pub inputs: Option<serde_json::Value>,
    pub outputs: Option<serde_json::Value>,
    pub urn: String,
    #[serde(rename = "type")]
    pub _type: String,
    pub id: Option<String>,
    pub provider: Option<String>,
    pub parent: Option<String>,
    pub source_position: Option<String>,
    pub modified: Option<String>,
    pub custom: Option<bool>,
    pub external: Option<bool>,
    pub created: Option<String>,
    pub dependencies: Option<Vec<String>>,
    pub additional_secret_outputs: Option<Vec<String>>,
    pub property_dependencies: Option<HashMap<String, Vec<String>>>,

    #[serde(flatten)]
    pub extra_values: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq, Hash)]
#[serde(rename_all = "kebab-case")]
pub enum Operation {
    Same,
    Read,
    Create,
    Update,
    Delete,
    Replace,
    Refresh,
    Import,
    CreateReplacement,
    DeleteReplaced,
    ReadReplacement,
    Discard,
    DiscardReplaced,
    RemovePendingReplace,
    ImportReplacement,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StackChangeStep {
    pub op: Operation,
    pub urn: String,
    pub provider: Option<String>,
    pub new_state: Option<ResourceState>,
    pub old_state: Option<ResourceState>,
    #[serde(flatten)]
    pub extra_values: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StackChangeSummary {
    pub steps: Vec<StackChangeStep>,
    pub change_summary: Option<HashMap<Operation, usize>>,
    pub duration: i64,
    #[serde(flatten)]
    pub extra_values: Option<serde_json::Value>,
}

pub struct PulumiProcessListener {
    pub preview_tx: Option<Sender<StackChangeSummary>>,
    pub event_tx: Sender<EngineEvent>,
}

#[derive(Debug, Clone, Default)]
pub struct StackRefreshOptions {
    pub skip_preview: Option<bool>,
    pub target: Option<Vec<String>>,
    pub target_dependents: Option<bool>,
    pub exclude: Option<Vec<String>>,
    pub exclude_dependents: Option<bool>,
    pub show_secrets: Option<bool>,
    pub continue_on_error: Option<bool>,
}

#[derive(Debug, Clone, Default)]
pub struct StackDestroyOptions {
    pub skip_preview: Option<bool>,
    pub refresh: Option<bool>,
    pub target: Option<Vec<String>>,
    pub target_dependents: Option<bool>,
    pub exclude: Option<Vec<String>>,
    pub exclude_dependents: Option<bool>,
    pub show_secrets: Option<bool>,
    pub continue_on_error: Option<bool>,
    pub exclude_protected: Option<bool>,
}

#[derive(Debug, Clone, Default)]
pub struct StackUpOptions {
    pub skip_preview: Option<bool>,
    pub preview: Option<bool>,
    pub refresh: Option<bool>,
    pub replace: Option<Vec<String>>,
    pub diff: Option<bool>,
    pub target: Option<Vec<String>>,
    pub target_dependents: Option<bool>,
    pub exclude: Option<Vec<String>>,
    pub exclude_dependents: Option<bool>,
    pub show_secrets: Option<bool>,
    pub continue_on_error: Option<bool>,
    pub expect_no_changes: Option<bool>,
    pub show_reads: Option<bool>,
    pub show_replacement_steps: Option<bool>,
}

#[derive(Debug, Clone, Default)]
pub struct StackPreviewOptions {
    pub refresh: Option<bool>,
    pub replace: Option<Vec<String>>,
    pub diff: Option<bool>,
    pub target: Option<Vec<String>>,
    pub target_dependents: Option<bool>,
    pub exclude: Option<Vec<String>>,
    pub exclude_dependents: Option<bool>,
    pub show_secrets: Option<bool>,
    pub continue_on_error: Option<bool>,
    pub expect_no_changes: Option<bool>,
    pub show_reads: Option<bool>,
    pub show_replacement_steps: Option<bool>,
}

#[async_trait::async_trait]
pub trait Stack: Sized {
    fn preview(&self, options: StackPreviewOptions) -> super::Result<StackChangeSummary>;
    async fn up(
        &self,
        options: StackUpOptions,
        listener: PulumiProcessListener,
    ) -> super::Result<()>;
    async fn refresh(
        &self,
        options: StackRefreshOptions,
        listener: PulumiProcessListener,
    ) -> super::Result<()>;
    async fn destroy(
        &self,
        options: StackDestroyOptions,
        listener: PulumiProcessListener,
    ) -> super::Result<()>;
}
