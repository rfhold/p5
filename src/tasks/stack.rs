use pulumi_automation::{
    event::EngineEvent,
    local::LocalStack,
    stack::{PulumiProcessListener, Stack, StackChangeSummary},
};
use tokio::sync::mpsc;
use tokio_util::task::TaskTracker;

use crate::{
    actions::{AppAction, StackAction},
    state::{OperationOptions, ProgramOperation},
};

use super::AppTask;

#[derive(Clone)]
pub enum StackTask {
    RunOperation(ProgramOperation, LocalStack, OperationOptions),
}

impl OperationOptions {
    pub fn skip_preview(mut self) -> Self {
        self.skip_preview = true;
        self
    }

    pub fn preview_only(mut self) -> Self {
        self.preview_only = true;
        self
    }
}

impl StackTask {
    #[tracing::instrument(skip(self, _task_tx, action_tx))]
    pub async fn run(
        &mut self,
        _task_tx: &mpsc::Sender<AppTask>,
        action_tx: &mpsc::Sender<AppAction>,
    ) -> crate::Result<()> {
        match self {
            StackTask::RunOperation(operation, local_stack, options) => {
                let (preview_tx, mut preview_rx) = mpsc::channel::<StackChangeSummary>(100);
                let (event_tx, mut event_rx) = mpsc::channel::<EngineEvent>(100);

                let tracker = TaskTracker::new();

                let preview_stack = local_stack.clone();
                let preview_action_tx = action_tx.clone();
                let preview_operation = operation.clone();
                tracker.spawn(async move {
                    while let Some(preview) = preview_rx.recv().await {
                        if let Err(err) = preview_action_tx.try_send(AppAction::StackAction(
                            StackAction::PersistChangeSummary(
                                preview_operation.clone(),
                                preview_stack.clone(),
                                preview,
                            ),
                        )) {
                            tracing::error!("Failed to send stack update change summary: {}", err);
                        }
                    }
                });

                let event_stack = local_stack.clone();
                let event_action_tx = action_tx.clone();
                let event_operation = operation.clone();
                tracker.spawn(async move {
                    while let Some(event) = event_rx.recv().await {
                        if let Err(err) = event_action_tx.try_send(AppAction::StackAction(
                            StackAction::PersistEvent(
                                event_operation.clone(),
                                event_stack.clone(),
                                event,
                            ),
                        )) {
                            tracing::error!("Failed to send stack update event: {}", err);
                        }
                    }
                });

                match operation {
                    ProgramOperation::Update if options.preview_only => {
                        if let Err(err) = local_stack
                            .preview_async(
                                pulumi_automation::stack::StackPreviewOptions {
                                    ..Default::default()
                                },
                                PulumiProcessListener {
                                    preview_tx: Some(preview_tx),
                                    event_tx,
                                },
                            )
                            .await
                        {
                            tracing::error!("Failed to preview stack: {:?}", err);
                        }
                    }
                    ProgramOperation::Update => {
                        let mut update_options = pulumi_automation::stack::StackUpOptions {
                            ..Default::default()
                        };

                        if options.skip_preview {
                            update_options.skip_preview = Some(true);
                        }
                        if let Err(err) = local_stack
                            .up(
                                update_options,
                                PulumiProcessListener {
                                    preview_tx: Some(preview_tx),
                                    event_tx,
                                },
                            )
                            .await
                        {
                            tracing::error!("Failed to preview stack: {:?}", err);
                        }
                    }
                    ProgramOperation::Refresh => {
                        let mut refresh_options = pulumi_automation::stack::StackRefreshOptions {
                            ..Default::default()
                        };

                        if options.skip_preview {
                            refresh_options.skip_preview = Some(true);
                        }
                        if let Err(err) = local_stack
                            .refresh(
                                refresh_options,
                                PulumiProcessListener {
                                    preview_tx: Some(preview_tx),
                                    event_tx,
                                },
                            )
                            .await
                        {
                            tracing::error!("Failed to refresh stack: {:?}", err);
                        }
                    }
                    ProgramOperation::Destroy => {
                        let mut destroy_options = pulumi_automation::stack::StackDestroyOptions {
                            ..Default::default()
                        };

                        if options.skip_preview {
                            destroy_options.skip_preview = Some(true);
                        }
                        if let Err(err) = local_stack
                            .destroy(
                                destroy_options,
                                PulumiProcessListener {
                                    preview_tx: Some(preview_tx),
                                    event_tx,
                                },
                            )
                            .await
                        {
                            tracing::error!("Failed to destroy stack: {:?}", err);
                        }
                    }
                }

                tracker.close();

                tracker.wait().await;

                action_tx.try_send(AppAction::StackAction(StackAction::PersistOperationDone(
                    operation.clone(),
                    local_stack.clone(),
                )))?;

                Ok(())
            }
        }
    }
}
