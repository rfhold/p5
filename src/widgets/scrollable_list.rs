use ratatui::{
    buffer::Buffer,
    layout::{Margin, Rect},
    style::Style,
    widgets::{
        Block, List, ListItem, ListState, Scrollbar, ScrollbarOrientation, ScrollbarState,
        StatefulWidget,
    },
};

#[derive(Default, Clone, Debug)]
pub struct ScrollableListState {
    pub list_state: ListState,
}

pub struct ScrollableList<'a> {
    items: Vec<ListItem<'a>>,
    block: Option<Block<'a>>,
}

impl<'a> ScrollableList<'a> {
    pub fn new(items: Vec<ListItem<'a>>) -> Self {
        Self { items, block: None }
    }

    pub fn block(mut self, block: Block<'a>) -> Self {
        self.block = Some(block);
        self
    }

    fn content_height(&self) -> usize {
        self.items.iter().map(|item| item.height()).sum::<usize>()
    }

    fn v_scroll_state(&self, list_state: &ListState, area: &Rect) -> Option<ScrollbarState> {
        let content_height = self.content_height();
        let area_height = area.height as usize;
        if content_height <= area_height {
            return None;
        }
        let position = self
            .items
            .iter()
            .take(list_state.selected().unwrap_or(0))
            .map(|item| item.height())
            .sum::<usize>();

        let overflow = content_height - area_height;
        let offset_position = position.saturating_sub(overflow);

        Some(ScrollbarState::new(overflow).position(offset_position))
    }

    fn v_scroll_area(&self, area: Rect) -> Rect {
        if let Some(_block) = &self.block {
            area.inner(Margin {
                horizontal: 0,
                vertical: 1,
            })
        } else {
            area
        }
    }
}

impl StatefulWidget for ScrollableList<'_> {
    type State = ScrollableListState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let v_scroll_area = self.v_scroll_area(area);
        let v_scroll_state = self.v_scroll_state(&state.list_state, &v_scroll_area);

        let list = List::new(self.items)
            .block(self.block.unwrap_or_default())
            .highlight_style(Style::default().fg(ratatui::style::Color::Yellow));

        StatefulWidget::render(list, area, buf, &mut state.list_state);

        if v_scroll_state.is_none() {
            return;
        }

        let v_scrollbar = Scrollbar::new(ScrollbarOrientation::VerticalRight);
        StatefulWidget::render(
            v_scrollbar,
            v_scroll_area,
            buf,
            &mut v_scroll_state.unwrap(),
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    #[test]
    fn test_basic_scrollable() {
        let list_items = vec![
            ListItem::new("Item 1"),
            ListItem::new("Item 2"),
            ListItem::new("Item 3"),
            ListItem::new("Item 4"),
        ];

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default(),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_overflow() {
        let list_items = vec![
            ListItem::new("Item 1"),
            ListItem::new("Item 2"),
            ListItem::new("Item 3"),
            ListItem::new(
                "Item 4, which is a bit longer than the others to test overflow. Turns out, it still needs to be a bit longer",
            ),
        ];

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default(),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_vertical_scroll() {
        let list_items = (1..=35)
            .map(|i| ListItem::new(format!("Item {}", i)))
            .collect::<Vec<_>>();

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default().with_selected(Some(30)),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_vertical_scroll_first_page() {
        let list_items = (1..=35)
            .map(|i| ListItem::new(format!("Item {}", i)))
            .collect::<Vec<_>>();

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default().with_selected(Some(17)),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_vertical_scroll_last_page() {
        let list_items = (1..=35)
            .map(|i| ListItem::new(format!("Item {}", i)))
            .collect::<Vec<_>>();

        let list = ScrollableList::new(list_items).block(
            Block::default()
                .title("Scrollable List")
                .borders(ratatui::widgets::Borders::ALL),
        );

        let mut state = ScrollableListState {
            list_state: ListState::default().with_selected(Some(34)),
        };

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(list, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
