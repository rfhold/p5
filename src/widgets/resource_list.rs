use pulumi_automation::stack::{Operation, ResourceState, StackChangeSummary};
use ratatui::{
    buffer::Buffer,
    layout::Rect,
    style::{Color, Modifier, Style},
    text::{Line, Span},
    widgets::{Block, ListItem, StatefulWidget},
};
use ratatui_macros::{line, span};

use super::scrollable_list::{ScrollableList, ScrollableListState};
use crate::state::ResourceOperationState;

#[derive(Debug, Clone)]
pub struct ResourceItem<'a> {
    pub urn: String,
    pub line: Line<'a>,
}

pub trait ResourceAdapter {
    fn get_items(&self) -> Vec<ResourceItem>;
}

#[derive(Default, Clone, Debug)]
pub struct ResourceListState {
    pub scrollable_state: ScrollableListState,
    pub excluded_resources: Vec<String>,
    pub target_resources: Vec<String>,
    pub replace_resources: Vec<String>,
}

pub struct ResourceList<'a, A: ResourceAdapter> {
    pub block: Block<'a>,
    pub adapter: A,
}

// Adapter implementations
pub struct CurrentStateAdapter<'a>(pub &'a Vec<ResourceState>);

impl<'a> ResourceAdapter for CurrentStateAdapter<'a> {
    fn get_items(&self) -> Vec<ResourceItem> {
        self.0
            .iter()
            .map(|resource| ResourceItem {
                urn: resource.urn.clone(),
                line: line!(
                    span!(Style::default().fg(Color::White); "{:25} ", resource.resource_type),
                    span!(Style::default().fg(Color::Cyan); "{:<}", resource.name()),
                ),
            })
            .collect()
    }
}

pub struct ChangeSummaryAdapter<'a>(pub &'a StackChangeSummary);

impl<'a> ResourceAdapter for ChangeSummaryAdapter<'a> {
    fn get_items(&self) -> Vec<ResourceItem> {
        self.0
            .steps
            .iter()
            .map(|step| {
                let state = if let Some(s) = step.new_state.as_ref() {
                    Some(s)
                } else {
                    step.old_state.as_ref()
                };

                if let Some(ref state) = state {
                    ResourceItem {
                        urn: state.urn.clone(),
                        line: line!(
                            operation_to_span(&step.op, Some(10)),
                            span!(Style::default().fg(Color::Gray); "{:<25} ", state.resource_type),
                            span!(Style::default(); "{:<}", state.name()),
                        ),
                    }
                } else {
                    ResourceItem {
                        urn: String::new(),
                        line: line!(
                            operation_to_span(&step.op, Some(10)),
                            span!(Style::default().fg(Color::Gray); "{:<25} ", "Unknown resource type"),
                            span!(Style::default(); "{:<}", "Unknown resource"),
                        ),
                    }
                }
            })
            .collect()
    }
}

pub struct OperationStateAdapter<'a>(pub &'a Vec<ResourceOperationState>);

impl<'a> ResourceAdapter for OperationStateAdapter<'a> {
    fn get_items(&self) -> Vec<ResourceItem> {
        self.0
            .iter()
            .map(|op_state| {
                let metadata = match op_state {
                    ResourceOperationState::InProgress { pre_event, .. } => &pre_event.metadata,
                    ResourceOperationState::Completed { out_event, .. } => &out_event.metadata,
                    ResourceOperationState::Failed { failed_event, .. } => &failed_event.metadata,
                };

                let pre_event = match op_state {
                    ResourceOperationState::InProgress { pre_event, .. } => pre_event,
                    ResourceOperationState::Completed { pre_event, .. } => pre_event,
                    ResourceOperationState::Failed { pre_event, .. } => pre_event,
                };

                let (dur_text, dur_style) = {
                    let (start, end) = match op_state {
                        ResourceOperationState::InProgress { start_time, .. } => {
                            (*start_time, chrono::Utc::now())
                        }
                        ResourceOperationState::Completed {
                            start_time,
                            end_time,
                            ..
                        } => (*start_time, *end_time),
                        ResourceOperationState::Failed {
                            start_time,
                            end_time,
                            ..
                        } => (*start_time, *end_time),
                    };

                    let duration_s = (end - start).num_seconds();
                    (
                        format!("{:.0}s", duration_s as f64),
                        Style::default().fg(match op_state {
                            ResourceOperationState::InProgress { .. } => Color::Gray,
                            ResourceOperationState::Completed { .. } => Color::DarkGray,
                            ResourceOperationState::Failed { .. } => Color::Red,
                        }),
                    )
                };

                let resource_type = metadata.resource_type.to_string();

                let name = metadata.name();

                ResourceItem {
                    urn: metadata.urn.clone(),
                    line: line!(
                        operation_to_span(&pre_event.metadata.op, Some(10)),
                        span!(dur_style; "{:<2} ", dur_text),
                        span!(Style::default().fg(Color::Gray); "{:.25} ", resource_type),
                        span!(Style::default(); "{:<}", name),
                    ),
                }
            })
            .collect()
    }
}

impl<'a, A: ResourceAdapter> StatefulWidget for ResourceList<'a, A> {
    type State = ResourceListState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let items = self.adapter.get_items();

        let list_items: Vec<ListItem> = items
            .iter()
            .map(|item| {
                let mut style = Style::default();

                // Apply selection styling (priority order matters)
                if state.excluded_resources.contains(&item.urn) {
                    style = style.fg(Color::Red).add_modifier(Modifier::CROSSED_OUT);
                } else if state.replace_resources.contains(&item.urn) {
                    style = style.fg(Color::Yellow).add_modifier(Modifier::ITALIC);
                } else if state.target_resources.contains(&item.urn) {
                    style = style.fg(Color::Green).add_modifier(Modifier::BOLD);
                }

                ListItem::new(item.line.clone()).style(style)
            })
            .collect();

        let scrollable_list = ScrollableList::new(list_items).block(self.block);
        scrollable_list.render(area, buf, &mut state.scrollable_state);
    }
}

impl<'a> ResourceList<'a, CurrentStateAdapter<'a>> {
    pub fn from_states(block: Block<'a>, states: &'a Vec<ResourceState>) -> Self {
        ResourceList {
            block,
            adapter: CurrentStateAdapter(states),
        }
    }
}

impl<'a> ResourceList<'a, ChangeSummaryAdapter<'a>> {
    pub fn from_summary(block: Block<'a>, summary: &'a StackChangeSummary) -> Self {
        ResourceList {
            block,
            adapter: ChangeSummaryAdapter(summary),
        }
    }
}

impl<'a> ResourceList<'a, OperationStateAdapter<'a>> {
    pub fn from_operations(block: Block<'a>, operations: &'a Vec<ResourceOperationState>) -> Self {
        ResourceList {
            block,
            adapter: OperationStateAdapter(operations),
        }
    }
}

fn operation_to_span(op: &Operation, width: Option<usize>) -> Span<'_> {
    let style = match op {
        Operation::Create => Style::default().fg(Color::Green),
        Operation::Update => Style::default().fg(Color::Yellow),
        Operation::Delete => Style::default().fg(Color::Red),
        Operation::Read => Style::default().fg(Color::Blue),
        Operation::Replace => Style::default().fg(Color::Magenta),
        _ => Style::default(),
    };

    match width {
        Some(w) => span!(style; "{:<w$}", op.to_string()),
        None => span!(style; "{}", op.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use pulumi_automation::{
        event::{ResOpFailedDetails, ResOutputsDetails, ResourcePreDetails},
        stack::{ResourceType, StackChangeStep},
    };
    use ratatui::{Terminal, backend::TestBackend, widgets::ListState};

    #[test]
    fn test_resources() {
        let items = vec![
            ResourceState {
                urn: "urn:pulumi:stack::project::pulumi:pulumi:Stack::project".to_string(),
                resource_type: ResourceType::Stack,
                ..Default::default()
            },
            ResourceState {
                urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket".to_string(),
                resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                ..Default::default()
            },
        ];

        let mut state = ResourceListState {
            scrollable_state: ScrollableListState {
                list_state: ListState::default(),
            },
            excluded_resources: vec![],
            target_resources: vec![],
            replace_resources: vec![],
        };

        let widget = ResourceList::from_states(
            Block::default()
                .title("Resources")
                .borders(ratatui::widgets::Borders::ALL),
            &items,
        );

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(widget, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_change_summary() {
        let change_summary = StackChangeSummary {
            steps: vec![
                StackChangeStep {
                    op: pulumi_automation::stack::Operation::Create,
                    old_state: None,
                    new_state: Some(ResourceState {
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
                StackChangeStep {
                    op: pulumi_automation::stack::Operation::Delete,
                    old_state: Some(ResourceState {
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::old-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        ..Default::default()
                    }),
                    new_state: None,
                    ..Default::default()
                },
            ],
            ..Default::default()
        };

        let mut state = ResourceListState {
            scrollable_state: ScrollableListState {
                list_state: ListState::default(),
            },
            excluded_resources: vec![],
            target_resources: vec![],
            replace_resources: vec![],
        };

        let widget = ResourceList::from_summary(
            Block::default()
                .title("Change Summary")
                .borders(ratatui::widgets::Borders::ALL),
            &change_summary,
        );

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(widget, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_operation_states() {
        let operation_states = vec![
            ResourceOperationState::InProgress {
                sequence: 1,
                start_time: chrono::Utc::now(),
                pre_event: ResourcePreDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Create,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
            },
            ResourceOperationState::Completed {
                sequence: 2,
                start_time: chrono::Utc::now(),
                end_time: chrono::Utc::now(),
                pre_event: ResourcePreDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Update,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
                out_event: ResOutputsDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Update,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
            },
            ResourceOperationState::Failed {
                sequence: 3,
                start_time: chrono::Utc::now(),
                end_time: chrono::Utc::now(),
                pre_event: ResourcePreDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Delete,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    planning: Some(false),
                    extra_values: None,
                },
                failed_event: ResOpFailedDetails {
                    metadata: pulumi_automation::event::StepEventMetadata {
                        op: pulumi_automation::stack::Operation::Delete,
                        urn: "urn:pulumi:stack::project::aws:s3/bucket:Bucket::my-bucket"
                            .to_string(),
                        resource_type: ResourceType::Other("aws:s3/bucket:Bucket".to_string()),
                        old: None,
                        new: None,
                        keys: None,
                        diffs: None,
                        detailed_diff: None,
                        logical: None,
                        provider: "aws".to_string(),
                    },
                    steps: 0,
                    status: 0,
                    extra_values: None,
                },
            },
        ];

        let mut state = ResourceListState {
            scrollable_state: ScrollableListState {
                list_state: ListState::default(),
            },
            excluded_resources: vec![],
            target_resources: vec![],
            replace_resources: vec![],
        };

        let widget = ResourceList::from_operations(
            Block::default()
                .title("Resource Operations")
                .borders(ratatui::widgets::Borders::ALL),
            &operation_states,
        );

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(widget, frame.area(), &mut state))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
