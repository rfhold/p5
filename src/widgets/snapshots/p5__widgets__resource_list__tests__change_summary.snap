---
source: src/widgets/resource_list.rs
assertion_line: 328
expression: terminal.backend()
---
"┌Change Summary────────────────────────────────────────────────────────────────┐"
"│Create    aws:s3/bucket:Bucket      my-bucket                                 │"
"│Delete    aws:s3/bucket:Bucket      old-bucket                                │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"└──────────────────────────────────────────────────────────────────────────────┘"
